Starting draft for git episode (#3)
UW-Madison-DataScience/R-gitlab-lesson: episodes/git.Rmd

Also the number lines for the removed lines are in the first column and the numbers for the new lines are in the second column.

## Exploring the history of our file changes

The Git pane has several other features for working with git in RStudio, the other one we will point out is the history part of the "Review Changes" window.
To open it directly you can click the click icon on the Git pane menu.
You can access it by clicking the "History" toggle in the Review Changes window, when it is already open.
The history pane now shows the "timeline" of commits we've added so far, authorship info, and we can click through them to see the diffs for each one. 

## Time to push it!

Finally we can sync these changes with GitLab.  Everything, besides the original setup, we have done so far has been on our local computer.
If we reload the GitLab project, nothing has changed.  The commits we made aren't there, nor is the script we created.
To get our changes to GitLab we have to "push" them.
In the git pane, we can click the green up arrow. This will then pop up a status as it transfers the changes.

Now in GitLab we can reload the page and we can see the script and our most recent changes.  

> In GitLab, try exploring the files by clicking on them. You can also see the commits and their diffs by clicking on the "X commits" (where X is the number of commits you've made) link
> in the right hand menu or by clicking "Code" and "Commits" buttons on the left hand menus.

Author: ```suggestion
In GitLab, try exploring the files by clicking on them. You can also see the commits and their diffs by clicking on the "X commits" (where X is the number of commits you've made) link
in the right hand "Project Information" menu or by clicking "Code" and "Commits" buttons on the left hand GitLab menu.
```

By default, it will expire in a month.  We will leave that setting for now and if you use GitLab after a month, you should go back and give it further permissions or create a new token.
Select "Owner" as the role.
You can leave all the options under "Select Scopes" unchecked, these would be needed if you were setting fine-grained permissions for your token.
Then click "Create project access token".

It will then show the project token in a green box on the screen but we won't be able to access it again so we need to copy it now (we can always setup others if needed later).
Click the copy button to copy the token. Copy this into a text file.

Now that we have set up the authentication, we can create the Project on our computer and setup the connection to GitLab, at the same time.
In GitLab, we need to click the blue "Code" button on the upper right of our GitLab project.
Then click the copy button to the right of the "Clone with HTTPS" option.

<!--- screenshot needed: the code menu and the copy button from the clone with HTTPS option highlighted"

> Then switch over to your RStudio window and click the new project button (Or click File > New Project).

Author: ```suggestion
Open RStudio and create a new project with the +shield icon or `File` > `New Project` and select the "Version Control" option in the new project wizard.
```

:::::::::::::::::::::::::::::::::::::: questions 

- How do I setup a repository in GitLab?
- How do I use git to record the history of my code in Rstudio?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Identify the location and functionality of Git integration within RStudio
- Create a repository in GitLab for version control
- Clone a GitLab repository as a project in RStudio

> - Make changes to the project, add, and commit them using Git

Reviewer: ```suggestion
- Make changes to the project, add, and commit them using git
```

::::::::::::::::::::::::::::::::::::::::::::::::

## Introduction

This lesson will pick up using the ratdat / portal dataset we used in the previous sessions.

To get started we need to setup the project we are going to be working with (both in R/RStudio and in GitLab).
You can setup the project first in either place but for this example we will start by creating the repository in GitLab.

## Setting up the GitLab Project

> First you need to navigate to the [UW-Madison GitLab instance][uw-gitlab] and login with your netid.

Reviewer: ```suggestion
First you need to navigate to the [UW-Madison GitLab instance][uw-gitlab] and login with your UW NetID.
```

Next we will create a new project, you can do so by clicking the "+" button then "new project/repository" icon in the upper-left hand of the screen OR you can click the blue "New Project" button on the right hand side of the screen.

<!--- screenshot needed: gitlab interface highlighting the + button and new project/repo and new project button -->

> On the "Create new project" page, choose "Create a blank project".

Reviewer: ```suggestion
On the "Create new project" page, choose "Create blank project".
```

Then click the "Create project button".

<!--- screenshots needed: the create new project page with the options given in the paragraph above highlighted-->

Now that we have created the project we can see the repository.  It created a nice README start, that is show at the bottom.

<!--- screenshot needed: view of the project repo after creation -->

::::::::: instructor 

You may want to give a short orentation to the different parts of the GitLab project webpage that seem relevant here:
For example:
- the file list
- issues/merge requests on the left pane
- project info on the right side (license, changelog etc)

::::::::::::::::::::

## Setting up the Project in RStudio

Next we will `clone` (copy) this project to our computer and make it a project in RStudio.
Before we can do so we need to set up verifying our identity with GitLab.

### GitLab Authentication 

We need to set up a special passkey called a Peronal Authenticalion Token (PAT) on GitLab and then we will enter it in RStudio when we need to connect to GitLab.

We need to note both our GitLab username and your PAT so we can use them when we set up the Project on our computer.

> In GitLab you can check your username by clicking on your profile icon in the upper left panel. It will say `@FIRSTNAME.LASTNAME` or something to that effect.

Reviewer: ```suggestion
In GitLab you can check your username by clicking on your profile icon in the upper left panel. It will say `@FIRSTNAME.LASTNAME` or your `@NetID`, or something to that effect.
```

Note down your username in a text file so you have it available when you need it later.

To setup the PAT, click the "Settings" option on the left hand side (bottom of the left pane), then choose "Access Tokens".

> On the "Project access tokens" page, click the "Add new token" button. Give the token a name specfific to the computer you are working from.

Reviewer: ```suggestion
On the "Project access tokens" page, click the "Add new token" button. Give the token a name specific to the computer you are working from. You can leave the Token description blank.
```

By default, it will expire in a month.  We will leave that setting for now and if you use GitLab after a month, you should go back and give it further permissions or create a new token.

> Select "Owner" as the role.

Reviewer: ```suggestion
Select "Owner" as the role.
Add a line about "Select scopes"? Explaining what it is and to leave all options unchecked?
```

Then click "Create project access token".

It will then show the project token in a green box on the screen but we won't be able to access it again so we need to copy it now (we can always setup others if needed later).
Click the copy button to copy the token. Copy this into a text file.

Now that we have set up the authentication, we can create the Project on our computer and setup the connection to GitLab, at the same time.

> In GitLab, we need to click the blue "Code" button on the upper right of our GitLab project.

Reviewer: ```suggestion
In GitLab, return to your new project directory `ratdat_lib_wksp` by clicking the link at the top of the page. Then click the blue "Code" button at the top of the directory information.
```

Then click the copy button to the right of the "Clone with HTTPS" option.

<!--- screenshot needed: the code menu and the copy button from the clone with HTTPS option highlighted"

Then switch over to your RStudio window and click the new project button (Or click File > New Project).

<!--- screenshot needed: highligting the new project button -->

> In the new project wizard, click the "Version Control" option.

Reviewer: Open RStudio and create a new project with the +shield icon or `File > New Project` and select the "Version Control" option.

Author: I think I made this change in the suggestion above but let me know if it still needs edits for what you were looking for.

On the next page, click the "Git" option.
Paste the HTTPS address you copied in the "Repository URL" option.
It will autofill in the "Project directory name" box with the slug from the URL when you click away from that box, but you can change it if you ever need to (which you will later).
You may want to change where the project is created in your file system, by default it will likely save it in your home folder/`~`.
We will change ours to Desktop so we can find it more easily by clicking the "Browse" button, and then choseing our `Desktop` folder and clicking "Open".
We can then click the "Create project" button.

<!--- screenshots needed of all the project wizard pages with the inputs described above -->

> Since we've never authenticated with GitLab before it will ask us for our username.

Reviewer: It did not ask me for this

Author: Have you authenticated with gitlab before?

The file pane in the lower right quadrant of RStudio will also show the...
    - `.gitignore` - a file that allows you to have untracked files in the folder on you computer that in the project folder - this was created when we setup the Rproject so git doesn't try to tack some R files that commonly are not tracked with git
    - `ratdat_lib_wksp.Rproj` - the Rproject folder (this is nice for reopening your project from the file folder) - this was created when we set up the Rproject since it didn't already exist
    - `README.md` - The documentation file that GitLab created for us when we set up the project

## Tracking files with git and RStudio

Let's go check out the "Git" file pane in the upper right quadrant!  It wasn't there until we cloned this git repository/project.
It currently lists the two files that were created when we set up the project, `ratdat_lib_wksp.Rproj` and `.gitignore`.
This is because these are new files that have not been tracked with git before.

To track these files in our repository we have to do two steps, first we stage the files and then we commit them to the history of the project/repo.

> To stage the files in Rstudio, we can click the button in the "Staged" column of the Git pane.

Reviewer: ```suggestion
To stage the files in Rstudio, we can click the checkboxes in the "Staged" column of the Git pane.
```

When we do this, the status of each file changes from "? ?" to "A" for "Added to the stage".

> Then to commit them, we can click the commit button in the menu of the Git pane.

Reviewer: ```suggestion
Then to commit them, we can click the "Commit pending changes" icon in the Git pane.
```
